File: frontend/src/services/video.ts
/**
 * 视频生成 API 服务
 */

import { request } from './api';

// ============================================================
// 类型定义
// ============================================================

export type VideoModel = 'sora-2-text-to-video' | 'sora-2-image-to-video' | 'sora-2-pro-storyboard';

export type VideoFrames = '10' | '15' | '25';

export type VideoAspectRatio = 'portrait' | 'landscape';

export type TaskStatus = 'pending' | 'processing' | 'success' | 'failed' | 'timeout';

export interface GenerateTextToVideoRequest {
  prompt: string;
  model?: VideoModel;
  n_frames?: VideoFrames;
  aspect_ratio?: VideoAspectRatio;
  remove_watermark?: boolean;
  wait_for_result?: boolean;
}

export interface GenerateImageToVideoRequest {
  prompt: string;
  image_url: string;
  model?: VideoModel;
  n_frames?: VideoFrames;
  aspect_ratio?: VideoAspectRatio;
  remove_watermark?: boolean;
  wait_for_result?: boolean;
}

export interface GenerateStoryboardVideoRequest {
  n_frames?: VideoFrames;
  storyboard_images?: string[];
  aspect_ratio?: VideoAspectRatio;
  wait_for_result?: boolean;
}

export interface GenerateVideoResponse {
  task_id: string;
  status: TaskStatus;
  video_url: string | null;
  duration_seconds: number;
  credits_consumed: number;
  cost_usd: number;
  cost_time_ms?: number;
}

export interface TaskStatusResponse {
  task_id: string;
  status: TaskStatus;
  video_url: string | null;
  fail_code?: string;
  fail_msg?: string;
}

export interface VideoModelInfo {
  model_id: string;
  description: string;
  requires_image_input: boolean;
  requires_prompt: boolean;
  supported_frames: string[];
  supports_watermark_removal: boolean;
  credits_per_second: number;
}

export interface VideoModelsResponse {
  models: VideoModelInfo[];
}

// ============================================================
// API 方法
// ============================================================

/**
 * 文本生成视频
 */
export async function generateTextToVideo(data: GenerateTextToVideoRequest): Promise<GenerateVideoResponse> {
  return request<GenerateVideoResponse>({
    method: 'POST',
    url: '/videos/generate/text-to-video',
    data,
  });
}

/**
 * 图片生成视频
 */
export async function generateImageToVideo(data: GenerateImageToVideoRequest): Promise<GenerateVideoResponse> {
  return request<GenerateVideoResponse>({
    method: 'POST',
    url: '/videos/generate/image-to-video',
    data,
  });
}

/**
 * 故事板视频生成
 */
export async function generateStoryboardVideo(data: GenerateStoryboardVideoRequest): Promise<GenerateVideoResponse> {
  return request<GenerateVideoResponse>({
    method: 'POST',
    url: '/videos/generate/storyboard',
    data,
  });
}

/**
 * 查询任务状态
 */
export async function queryVideoTaskStatus(taskId: string): Promise<TaskStatusResponse> {
  return request<TaskStatusResponse>({
    method: 'GET',
    url: `/videos/tasks/${taskId}`,
  });
}

/**
 * 获取可用的视频模型列表
 */
export async function getVideoModels(): Promise<VideoModelsResponse> {
  return request<VideoModelsResponse>({
    method: 'GET',
    url: '/videos/models',
  });
}

// ============================================================
// 轮询工具
// ============================================================

export interface PollOptions {
  /** 轮询间隔（毫秒），默认 5000 (视频生成较慢) */
  interval?: number;
  /** 最大等待时间（毫秒），默认 600000 (10分钟) */
  maxWait?: number;
  /** 状态更新回调 */
  onStatusChange?: (status: TaskStatus) => void;
}

/**
 * 轮询视频任务直到完成
 */
export async function pollVideoTaskUntilDone(
  taskId: string,
  options: PollOptions = {}
): Promise<TaskStatusResponse> {
  const { interval = 5000, maxWait = 600000, onStatusChange } = options;
  const startTime = Date.now();

  while (true) {
    const result = await queryVideoTaskStatus(taskId);

    onStatusChange?.(result.status);

    if (result.status === 'success' || result.status === 'failed' || result.status === 'timeout') {
      return result;
    }

    if (Date.now() - startTime > maxWait) {
      throw new Error('视频生成等待超时');
    }

    await new Promise((resolve) => setTimeout(resolve, interval));
  }
}

// ============================================================
// 便捷方法
// ============================================================

/**
 * 生成视频并等待结果（自动处理轮询）
 */
export async function generateTextToVideoAndWait(
  data: Omit<GenerateTextToVideoRequest, 'wait_for_result'>,
  options?: PollOptions
): Promise<GenerateVideoResponse> {
  // 先异步创建任务
  const response = await generateTextToVideo({ ...data, wait_for_result: false });

  // 轮询直到完成
  const result = await pollVideoTaskUntilDone(response.task_id, options);

  return {
    ...response,
    status: result.status,
    video_url: result.video_url,
  };
}

/**
 * 图生视频并等待结果（自动处理轮询）
 */
export async function generateImageToVideoAndWait(
  data: Omit<GenerateImageToVideoRequest, 'wait_for_result'>,
  options?: PollOptions
): Promise<GenerateVideoResponse> {
  // 先异步创建任务
  const response = await generateImageToVideo({ ...data, wait_for_result: false });

  // 轮询直到完成
  const result = await pollVideoTaskUntilDone(response.task_id, options);

  return {
    ...response,
    status: result.status,
    video_url: result.video_url,
  };
}

// ============================================================
// 模型配置（前端展示用）
// ============================================================

export const VIDEO_MODELS = [
  {
    id: 'sora-2-text-to-video' as VideoModel,
    name: 'Sora 2 Text-to-Video',
    description: '文本生成视频',
    creditsPerSecond: 3,
    requiresImage: false,
    requiresPrompt: true,
    supportedDurations: [10, 15],
    supportsWatermarkRemoval: true,
  },
  {
    id: 'sora-2-image-to-video' as VideoModel,
    name: 'Sora 2 Image-to-Video',
    description: '图片生成视频',
    creditsPerSecond: 3,
    requiresImage: true,
    requiresPrompt: true,
    supportedDurations: [10, 15],
    supportsWatermarkRemoval: true,
  },
  {
    id: 'sora-2-pro-storyboard' as VideoModel,
    name: 'Sora 2 Pro Storyboard',
    description: '专业故事板视频',
    creditsPerSecond: 15,
    requiresImage: false,
    requiresPrompt: false,
    supportedDurations: [10, 15, 25],
    supportsWatermarkRemoval: false,
  },
];

export const VIDEO_DURATIONS = [
  { value: '10' as VideoFrames, label: '10秒', credits: 30 }, // Text/Image-to-Video 基准价格
  { value: '15' as VideoFrames, label: '15秒', credits: 45 }, // Text/Image-to-Video 基准价格
  { value: '25' as VideoFrames, label: '25秒', credits: 270, note: '仅故事板' }, // Pro Storyboard 价格
];

export const VIDEO_ASPECT_RATIOS = [
  { value: 'landscape' as VideoAspectRatio, label: '横屏' },
  { value: 'portrait' as VideoAspectRatio, label: '竖屏' },
];

File: backend/api/routes/video.py
"""
视频生成路由

提供视频生成、任务查询接口。
"""

from fastapi import APIRouter, Depends

from api.deps import CurrentUser, Database
from schemas.video import (
    GenerateTextToVideoRequest,
    GenerateImageToVideoRequest,
    GenerateStoryboardVideoRequest,
    GenerateVideoResponse,
    TaskStatusResponse,
    VideoModelsResponse,
    VideoModelInfo,
    TaskStatus,
)
from services.video_service import VideoService

router = APIRouter(prefix="/videos", tags=["视频生成"])


def get_video_service(db: Database) -> VideoService:
    """获取视频服务实例"""
    return VideoService(db)


@router.post("/generate/text-to-video", response_model=GenerateVideoResponse, summary="文本生成视频")
async def generate_text_to_video(
    request: GenerateTextToVideoRequest,
    current_user: CurrentUser,
    service: VideoService = Depends(get_video_service),
):
    """
    文生视频接口

    支持模型：
    - sora-2-text-to-video: 文本生成视频（40积分/10秒，60积分/15秒）
    - sora-2-pro-storyboard: 专业故事板（100积分/10秒，150积分/15秒，250积分/25秒）

    如果 wait_for_result=False，将立即返回 task_id，需要轮询 /tasks/{id} 获取结果。
    """
    result = await service.generate_text_to_video(
        user_id=current_user["id"],
        prompt=request.prompt,
        model=request.model.value,
        n_frames=request.n_frames.value,
        aspect_ratio=request.aspect_ratio.value,
        remove_watermark=request.remove_watermark,
        wait_for_result=request.wait_for_result,
    )

    return GenerateVideoResponse(
        task_id=result["task_id"],
        status=TaskStatus(result["status"]),
        video_url=result.get("video_url"),
        duration_seconds=result.get("duration_seconds", 0),
        credits_consumed=result.get("credits_consumed", 0),
        cost_usd=result.get("cost_usd", 0.0),
        cost_time_ms=result.get("cost_time_ms"),
    )


@router.post("/generate/image-to-video", response_model=GenerateVideoResponse, summary="图片生成视频")
async def generate_image_to_video(
    request: GenerateImageToVideoRequest,
    current_user: CurrentUser,
    service: VideoService = Depends(get_video_service),
):
    """
    图生视频接口

    使用 sora-2-image-to-video 模型，需要提供首帧图片。
    积分消耗：40积分/10秒，60积分/15秒
    """
    result = await service.generate_image_to_video(
        user_id=current_user["id"],
        prompt=request.prompt,
        image_url=request.image_url,
        model=request.model.value,
        n_frames=request.n_frames.value,
        aspect_ratio=request.aspect_ratio.value,
        remove_watermark=request.remove_watermark,
        wait_for_result=request.wait_for_result,
    )

    return GenerateVideoResponse(
        task_id=result["task_id"],
        status=TaskStatus(result["status"]),
        video_url=result.get("video_url"),
        duration_seconds=result.get("duration_seconds", 0),
        credits_consumed=result.get("credits_consumed", 0),
        cost_usd=result.get("cost_usd", 0.0),
        cost_time_ms=result.get("cost_time_ms"),
    )


@router.post("/generate/storyboard", response_model=GenerateVideoResponse, summary="故事板视频生成")
async def generate_storyboard_video(
    request: GenerateStoryboardVideoRequest,
    current_user: CurrentUser,
    service: VideoService = Depends(get_video_service),
):
    """
    故事板视频生成接口

    使用 sora-2-pro-storyboard 模型，支持最长25秒视频。
    积分消耗：100积分/10秒，150积分/15秒，250积分/25秒
    """
    result = await service.generate_storyboard_video(
        user_id=current_user["id"],
        model="sora-2-pro-storyboard",
        n_frames=request.n_frames.value,
        storyboard_images=request.storyboard_images,
        aspect_ratio=request.aspect_ratio.value,
        wait_for_result=request.wait_for_result,
    )

    return GenerateVideoResponse(
        task_id=result["task_id"],
        status=TaskStatus(result["status"]),
        video_url=result.get("video_url"),
        duration_seconds=result.get("duration_seconds", 0),
        credits_consumed=result.get("credits_consumed", 0),
        cost_usd=result.get("cost_usd", 0.0),
        cost_time_ms=result.get("cost_time_ms"),
    )


@router.get("/tasks/{task_id}", response_model=TaskStatusResponse, summary="查询任务状态")
async def query_task(
    task_id: str,
    current_user: CurrentUser,
    service: VideoService = Depends(get_video_service),
):
    """
    查询视频生成任务状态

    用于轮询异步任务的完成状态。
    """
    result = await service.query_task(task_id)

    return TaskStatusResponse(
        task_id=result["task_id"],
        status=TaskStatus(result["status"]),
        video_url=result.get("video_url"),
        fail_code=result.get("fail_code"),
        fail_msg=result.get("fail_msg"),
    )


@router.get("/models", response_model=VideoModelsResponse, summary="获取可用模型")
async def get_models(
    current_user: CurrentUser,
    service: VideoService = Depends(get_video_service),
):
    """
    获取可用的视频生成模型列表
    """
    models = service.get_available_models()

    return VideoModelsResponse(
        models=[VideoModelInfo(**m) for m in models]
    )
